### DISCUSSION TITLE
事件傳遞機制

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#130] [其他問題] [answered]
不太理解後面應該怎麼解釋

首先，透過 e.preventDefault() 我們阻止了超連結的預設行為，所以怎麼點都點不到 google，我們可以在 console 清楚地看到點擊超連結，事件傳遞的過程。先捕捉後冒泡。
```html
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="stylesheet" href="./style.css">
  
  <title>test</title>
</head>
<body>
  <form>
    <a class ='link' href="https://www.google.com/">
      link to google
    </a>
  </form>

  <script>
    const element = document.querySelector('.link')
    element.addEventListener('click', function(e) {
      e.preventDefault()
    })

    addEvent('.link')
    function addEvent(className) {
      document.querySelector(className)
      .addEventListener('click', function(e) {
        console.log(className, 'capture')
      },true)

      document.querySelector(className)
      .addEventListener('click', function() {
        console.log(className, 'bubble')
      },false)
    }
  </script>

</body>
</html>
```
接著，做一些微幅的調整，**問題就來了**

現在，我們在捕捉階段，裝上 e.stopPropagation()

```js
    function addEvent(className) {
      document.querySelector(className)
      .addEventListener('click', function(e) {
        e.stopPropagation()
        console.log(className, 'capture')
      },true)
```
為了能清楚看到狀況，先將超連結部分更改為`href="#"`
這個時候，因為 stopPropagation( ) 所以事件傳遞鏈斷開了，事件傳不到冒泡階段。

問題一、console 會顯示 link capture 是因為我點擊的時後，他雖然斷開後面的，但原本的事件還是會執行，而原本事件執行的這個階段叫做 target phase 對嗎?

問題二、現在將超連結部分改回原本的`href="https://www.google.com/"`，可以順利地連到 google.com。為甚麼 e.preventDefault( ) 失靈了? 和傳遞鏈斷開有關嗎 ? 
我推測是因為 click 事件現在都放在同一個 element 上面，所以當我將傳遞鏈斷開，原本 click 狀況要做的阻止預設函數就根本無法執行，我的推測對嗎 ? 

感謝大家的幫忙~

**Comment by allenliaooooooo** [comment]
> 問題一、console 會顯示 link capture 是因為我點擊的時後，他雖然斷開後面的，但原本的事件還是會執行，而原本事件執行的這個階段叫做 target phase 對嗎?

對~
> 問題二、現在將超連結部分改回原本的`href="https://www.google.com/"`，可以順利地連到 google.com。為甚麼 e.preventDefault( ) 失靈了? 和傳遞鏈斷開有關嗎 ?
> 我推測是因為 click 事件現在都放在同一個 element 上面，所以當我將傳遞鏈斷開，原本 click 狀況要做的阻止預設函數就根本無法執行，我的推測對嗎 ?

~~因為你加了三個 event listener 在同一個元素上面
而你只有 preventDefault 第一個~~
我上面的回答想錯了

~~剛剛試了一下發現
如果在 `preventDefault` 那個 event listener 加上第三個參數
`true` 時 `preventDefault` 有效
`false` 時 `preventDefault`  無效~~

~~所以我猜應該是沒傳入第三個參數的話就要整個事件傳遞回到 root 才會執行 function？~~
上面又想錯了，是太久沒寫 JavaScript 的後遺症嗎QQ
因為沒傳入第三個參數，所以預設是 false
第二個 event listener 用了 stopPropagation
所以冒泡階段的 event listener 都接收不到了
所以你的推測是對的

**Reply by s103071049** [reply]
感謝，好像對事件傳遞有比較了解。但有點疑惑，如果事件傳遞沒有完整跑完，裡面傳遞的事件還有辦法正常的執行嗎?

**Reply by allenliaooooooo** [reply]
> 裡面傳遞的事件還有辦法正常的執行嗎?

不太清楚這句話的意思
傳遞的事件是指例如點擊之後會執行的 function 嗎？
如果是的話，要看你把 listener 加在什麼地方
只要事件有傳遞到那裏，就算沒有冒泡回到最上層，還是會執行 function

**Reply by s103071049** [reply]
喔喔，好，我理解了。沒錯~
是這個意思!!
非常感謝~

**Comment by aszx87410** [comment, marked as answer]
``` js
const element = document.querySelector('.link')
// listener1
element.addEventListener('click', function(e) {
  e.preventDefault()
})

addEvent('.link')
function addEvent(className) {
  // listener2
  document.querySelector(className)
  .addEventListener('click', function(e) {
    console.log(className, 'capture')
  },true)

  // listener3
  document.querySelector(className)
  .addEventListener('click', function() {
    console.log(className, 'bubble')
  },false)
}
```

你依序加了三個 listener，第一個沒傳參數所以是冒泡階段的 listener，第二個捕獲階段，第三個冒泡階段

因為你的 listener 是加在 .link 上面，按的也是 .link，所以跑到這三個 listener 的時候都是 target 階段了

而因為是 target 階段，所以執行順序應該是按照加上 listener 的順序， listener1 => listener2 => listener3

但如同第五期頻道有同學[提到](https://juejin.cn/post/6965682915141386254)的，現在新的實作是就算 target 階段也會按照捕獲冒泡順序來執行，所以執行順序是：listener2 => listener1 => listener3（先捕獲，然後兩個冒泡照加上去的順序）

然後 `e.stopPropogation` 並不是「停止事件執行」，他只是停止事件繼續傳遞下去，而你的 listener1 已經被觸發了，所以並不是跑到那一行底下就都不跑了，因為 stopPropogation 並沒有這種功能

然後因為阻止事件傳遞了，所以 listener2 跟 listener3 都不會被執行到

**Reply by s103071049** [reply]
元素在執行預設動作的時候，為了對這個元素執行動作，會加監聽器，一旦事件被觸發就會執行後面的 function。監聽器會去攔截事件，針對攔截到的事件內容去做處理。target 只是在闡述事件傳遞的階段，從window 往下傳到目標物target，在從 target 往上傳到 window。所以，元素在執行預設動作時，事件傳到元素自己，自己這個階段叫做 target，事件經過 target 被監聽器攔劫，然後經過我們處理(寫回呼函式)，再執行繼續傳遞下去。看完影片多次後，還有個小小問題，事件傳遞的過程和事件傳遞的內容物是沒有關係的，但一定要經過完整的捕捉再到冒泡，window 下去再上來，傳遞的事件才會成功的被完成。這樣的理解對嗎?如果事件傳遞不完整無法冒泡到 window ，裡面的事件是不是也無法順利執行?

**Reply by outshaker** [reply]
@s103071049 那個，小小抓一下語病。target phase 和 target  是兩個不同的意思，target phase是指事件傳遞的階段，target 是事件要傳送到的對象，雖然按照上下文可以判斷你的意思比較不建議在這裡省略用語。

元素的預設動作是不是和事件監聽是同樣的相同機制？
我不確定，但放在一起討論的話會需要確認更多事情，我會比較傾向只針對「事件監聽」本身討論就好。

> 事件傳遞的過程和事件傳遞的內容物是沒有關係的，但一定要經過完整的捕捉再到冒泡，window 下去再上來，傳遞的事件才會成功的被完成。

「事件是否完成」其實是很難界定的事情，或者說，事件從來就不是要被完成的東西或指令，事件只是一個「信號」。

就像玩波浪舞的時候，那個在舞台前面拿著旗子跑的人就是「事件」，我們可以決定旗子(事件) 經過的時候要不要站起來，正常情況下這不會影響事件的傳遞，旗子照樣子跑它自己的。唯一例外是 stopPropagation() 它會讓整個旗子停下沒辦法跑完全程。

可能很多人(包括我自己) 一開始對事件的認知比較像是把信件 (event) 寄給收件人 (target) 的概念，但其實在事件傳遞的機制下，這一切比較像是大家在跳波浪舞的情境。旗子在揮，我們自己決定旗子經過的時候要不要站起來，這邊對應的程式操作就是增加監聽器(listener)。

如果你說旗子 (事件) 必須從頭跑到尾，然後再尾跑回頭這一整段過程走完才算是「事件完成」的話，確實也可以。但通常情況是只有幾個人會對經過的旗子 (事件) 有反應，所以著眼的點會是那些應該要對旗子 (事件) 有反應的人是不是有反應？如果再把事件代理機制加進來討論的話，情況又會更複雜，但大部分都還是討論事件是否有觸發原本預期要有的動作？

> 事件傳遞的過程和事件傳遞的內容物是沒有關係的

這邊沒辦法知道你的沒有關係是哪一種關係，但是稍微講一下「事件傳遞的內容物」是 event，裡面其實放了很多跟事件傳遞有關的狀態參數，所以其實還是有關係的。

所以可能要確定你指的「事件傳遞的內容物」是什麼才有辦法繼續討論

> 如果事件傳遞不完整無法冒泡到 window ，裡面的事件是不是也無法順利執行?

因為 addEventListener(eventName, handler) 這預設是冒泡階段才會觸發監聽器，你如果改成 addEventListener(eventName, handler, true) 就會變成捕獲階段觸發監聽器，這樣就可以確保在旗子來的時候(捕獲) 做動作。

會影響事件流的唯一狀況是前面有人使用 stopPropagation()，那樣的話，事件旗子不會到他這邊自然也不會觸發監聽器了。

**Reply by s103071049** [reply]
感謝大家的幫忙，對於事件傳遞有更理解!!謝謝~
